### PR TITLE
Add dao methods & ran checkstyle

### PR DESCRIPTION
**Pull request by zoercai**

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 1]
```diff
@@ -1384,49 +1384,52 @@ public ReadChangeStream withDeserializer(Deserializer deserializer) {
...
       pipelineInitializer.initialize(
           databaseAdminClient,
           partitionMetadataDao,
           databaseId,
           getInclusiveStartAt(),
           getExclusiveEndAt());
 
-      // FIXME: This should come from the generated table name
-      final String tableName = "tableName";
       final DaoFactory daoFactory = new DaoFactory(getChangeStreamName());
```
Comment (zoercai) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1427]: this creates a PartitionMetadataDao, which has already been created on line 1413

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 2]
```diff
@@ -1,12 +1,34 @@
+/*
```
Comment (zoercai) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:1]: Most of the changes are in here!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 3]
```diff
@@ -24,54 +46,85 @@
...
+                    + " CONTAINS(@partition)"
+                    + " AND "
+                    + COLUMN_STATE
+                    + " IN (@states)")
+            .bind("table")
+            .to(tableName)
+            .bind("partition")
+            .to(partitionToken)
+            .bind("states")
+            .toStringArray(states.stream().map(Enum::toString).collect(Collectors.toList()))
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:73]: Does this work without an `UNNEST`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 4]
```diff
@@ -24,54 +46,85 @@
...
+                    + " IN ("
+                    + " SELECT "
+                    + COLUMN_PARENT_TOKEN
+                    + " FROM @table"
+                    + " WHERE "
+                    + COLUMN_PARTITION_TOKEN
+                    + " = "
+                    + "@partition"
+                    + ")")
+            .bind("table")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:97]: Same as question as above

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 5]
```diff
@@ -24,54 +46,85 @@
...
     return databaseClient
         .readWriteTransaction()
-        .run(transaction -> {
-          // FIXME: table name should be given in the constructor instead of null
-          final InTransactionContext dao = new InTransactionContext(null, transaction);
-          return callable.apply(dao);
-        });
+        .run(
+            transaction -> {
+              final InTransactionContext dao = new InTransactionContext(tableName, transaction);
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:125]: nit: Sorry, I thing I misnamed this variable, could we change it to something like `transactionContext`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 6]
```diff
@@ -84,33 +137,46 @@ public InTransactionContext(String tableName, TransactionContext transaction) {
...
     }
 
-    // TODO: Implement
     public long countPartitionsInStates(
-        List<String> partitionTokens,
-        List<PartitionMetadata.State> states
-    ) {
-      throw new UnsupportedOperationException("Unimplemented");
+        List<String> partitionTokens, List<PartitionMetadata.State> states) {
+      final ResultSet resultSet =
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:157]: This `ResultSet resultSet` should be in a `try with resources` block

**Review by thiagotnunes — CHANGES_REQUESTED** [review]
Requesting changes just because of the ResultSet outside of the `try with resources` block.

Could you run some manual (integration) testing before merging this? I am wondering if some of the queries will not work (per comments below), but I might be wrong.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 7]
```diff
@@ -24,54 +46,85 @@
...
+        Statement.newBuilder(
+                "SELECT COUNT(*)"
+                    + "FROM @table"
+                    + " WHERE "
+                    + COLUMN_PARENT_TOKEN
+                    + " CONTAINS(@partition)"
+                    + " AND "
+                    + COLUMN_STATE
+                    + " IN (@states)")
+            .bind("table")
```
[the remark this answers is not in the document]
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:68]: Tested it out and no 😞 I've changed these to use string concat

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java (hunk 2 above)
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:1]: have removed the checkstyle changes

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 1 above)
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1423]: added a todo

**Review by thiagotnunes — APPROVED** [review]

**Pull request merged**